basculechecks/metricvalidator.go

```diff
@@ -0,0 +1,183 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package basculechecks
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/xmidt-org/bascule"
+	"github.com/xmidt-org/webpa-common/logging"
+)
+
+var (
+	ErrNilChecker  = errors.New("capability checker cannot be nil")
+	ErrNilMeasures = errors.New("capability check measures cannot be nil")
+)
+
+var defaultLogger = log.NewNopLogger()
+
+type CapabilitiesChecker interface {
+	Check(auth bascule.Authentication) (string, error)
+}
+
+type metricValidator struct {
+	c         CapabilitiesChecker
+	measures  *AuthCapabilityCheckMeasures
+	endpoints []*regexp.Regexp
+}
+
+func (m *metricValidator) CreateValidator(errorOut bool) bascule.ValidatorFunc {
+	return func(ctx context.Context, _ bascule.Token) error {
+		// if we're not supposed to error out, the outcome should be accepted on failure
+		failureOutcome := AcceptedOutcome
+		if errorOut {
+			// if we actually error out, the outcome is the request being rejected
+			failureOutcome = RejectedOutcome
+		}
+
+		auth, ok := bascule.FromContext(ctx)
+		if !ok {
+			m.measures.CapabilityCheckOutcome.With(OutcomeLabel, failureOutcome, ReasonLabel, TokenMissing, ClientIDLabel, "", PartnerIDLabel, "", EndpointLabel, "").Add(1)
+			if errorOut {
+				return ErrNoAuth
+			}
+			return nil
+		}
+
+		client, partnerID, endpoint, reason, err := m.prepMetrics(auth)
+		labels := []string{ClientIDLabel, client, PartnerIDLabel, partnerID, EndpointLabel, endpoint}
+		if err != nil {
+			labels = append(labels, OutcomeLabel, failureOutcome, ReasonLabel, reason)
+			m.measures.CapabilityCheckOutcome.With(labels...).Add(1)
+			if errorOut {
+				return err
+			}
+			return nil
+		}
+
+		reason, err = m.c.Check(auth)
+		if err != nil {
+			labels = append(labels, OutcomeLabel, failureOutcome, ReasonLabel, reason)
+			m.measures.CapabilityCheckOutcome.With(labels...).Add(1)
+			if errorOut {
+				return err
+			}
+			return nil
+		}
+
+		labels = append(labels, OutcomeLabel, AcceptedOutcome, ReasonLabel, "")
+		m.measures.CapabilityCheckOutcome.With(labels...).Add(1)
+		return nil
+	}
+}
+
+func NewMetricCapabilityChecker(c CapabilitiesChecker, m *AuthCapabilityCheckMeasures, endpoints []*regexp.Regexp) (*metricValidator, error) {
+	if c == nil {
+		return nil, ErrNilChecker
+	}
+	if m == nil {
+		return nil, ErrNilMeasures
+	}
+	mc := metricValidator{
+		measures:  m,
+		c:         c,
+		endpoints: endpoints,
+	}
+	return &mc, nil
+}
+
+func NewMetricCapabilityCheckerFromStrings(c CapabilitiesChecker, m *AuthCapabilityCheckMeasures, endpoints []string, logger log.Logger) (*metricValidator, error) {
```

Comment:
This function is an example of what I was getting at earlier.  It's signature is kinda lengthy, and over time it may grow with even more parameters.  If you just export `MetricValidator` and export the relevant fields, it'll save you having to maintain this function.

As long as you write all your code internally to the `CapabilitiesChecker` interface, you've accomplished good encapsulation.  Additionally, code outside `bascule` can be implemented in terms of `CapabilitiesChecker`.  So, the only code that needs to be aware of the specific struct and its fields is the point where its created.